### PR TITLE
Port to concurrent hash tables

### PR DESCRIPTION
**Pull request by valarauca**
This PR build's in [Ticki's](https://github.com/ticki) [`CHashMap`](https://crates.io/crates/chashmap) library.

The goal is to move away from global `RwLocks` to per bucket locks.

---

The PR is inflated because I ran `rustfmt`. I saw the `toml` and I thought that was protocol. I can attempt to revert that if you want(?)

---

On my lap this appears to be within margin of error for performance testing. But also none of benches appear to be doing multi-threading so idk.

### CODE REVIEW

File: src/vec.rs.bk [hunk 1]
```diff
@@ -0,0 +1,420 @@
+// Copyright 2014 The Prometheus Authors
```
Comment (siddontang) [inline src/vec.rs.bk]: please remove this backup file

**Comment by siddontang** [comment]
Thanks @valarauca 

Seem that you use another rustfmt version to format the code, we should use rustfmt nightly 0.2.1 version here.

Do you compare the benchmark with the concurrent hash map and origin implementation.

Maybe you can benchmark with #121 too.

**Comment by valarauca** [comment]
Yes I'll follow on that up. Had to travel this weekend.

**Comment by valarauca** [comment]
Sorry for punting on this. I'll resubmit at a later date when its more complete.

**Comment by siddontang** [comment]
Thanks @valarauca 

Maybe we can use SpinLock at first. You can have a try :-)

**Pull request closed without merging**